capstone/src/test/java/com/google/sps/servlets/CheckNewUserServletTest.java

```diff
@@ -0,0 +1,136 @@
+package com.google.sps.servlets;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.assertFalse;
+import static org.mockito.Mockito.when;
+
+import com.google.sps.Objects.User;
+import com.google.sps.Objects.Badge;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalUserServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.ArrayList;
+
+/**
+ * Unit tests for {@link CheckNewUserServlet}.
+ */
+ @RunWith(JUnit4.class)
+public class CheckNewUserServletTest {
```

Comment:
Okay. Can tests for toEntity and fromEntity go in CheckNewUserServletTest.java or should they go in a new class specifically to test methods in User.java?

Reply:
Since the toEntity and fromEntity methods are specific to the User class, they should be tested independently in a UserTest

Reply:
> This test would still pass if a user didn't add the new field to EITHER, so I would see if you can also come up with a testing strategy that would cause a test to fail if a user added a new field to the object but didn't add it to `toEntity/fromEntity`

In fromEntity, a User object is made using the constructor. If we added a field to the object and added that field to the constructor, but didn't update the fromEntity method, then fromEntity would throw an error. This would force fromEntity to be updated when the object is updated, which in turn would cause the round-trip test to fail if toEntity wasn't also updated.